lib/Dialect/TTIR/Transforms/Passes.cpp

```diff
@@ -1453,4 +1455,121 @@ class TTIRLoadSystemDesc
   }
 };
 
+class TTIRGatherPatternMatch
+    : public ttir::impl::TTIRGatherPatternMatchBase<TTIRGatherPatternMatch> {
+public:
+  using ttir::impl::TTIRGatherPatternMatchBase<
+      TTIRGatherPatternMatch>::TTIRGatherPatternMatchBase;
+  void runOnOperation() final {
+    ModuleOp module = getOperation();
+    IRRewriter rewriter(&getContext());
+    Operation *gatherOp = nullptr;
+    bool replaceWithEmbedding = true;
+    bool reduceWeightTensorDim = false;
+
+    module->walk([&](GatherOp op) {
+      gatherOp = op;
+      // if conditions are met, replace gather with embedding
+      auto outputType = mlir::cast<RankedTensorType>(op.getResult().getType());
+      auto shape = outputType.getShape();
+      auto hiddenDim = shape[1];
+      auto startIndices =
+          op.getStartIndices(); // start indices of the gather op
+      auto startIndicesType =
+          mlir::cast<RankedTensorType>(startIndices.getType());
+      auto sliceSizes = op.getSliceSizes(); // slice sizes of the gather op
+      auto offsetDims = op.getOffsetDims();
+      auto collapsedSliceDims =
+          op.getCollapsedSliceDims(); // collapsed slice dims of the gather op
+
+      if (shape.size() > 1) {
+        if (sliceSizes[0] != 1 && sliceSizes[1] != hiddenDim) {
+          replaceWithEmbedding = false;
+          llvm::outs() << "Did not satisfy 0" << "\n";
+        }
+      }
+      if (offsetDims.size() != 1 &&
+          std::vector<int64_t>(offsetDims.begin(), offsetDims.end()) !=
+              std::vector<int64_t>{2}) {
+        replaceWithEmbedding = false;
+        llvm::outs() << "Did not satisfy 1" << "\n";
+      }
+      if (collapsedSliceDims.size() != 1 &&
+          std::vector<int64_t>(collapsedSliceDims.begin(),
+                               collapsedSliceDims.end()) !=
+              std::vector<int64_t>{1}) {
+        replaceWithEmbedding = false;
+        llvm::outs() << "Did not satisfy 2" << "\n";
+      }
+      if (shape.size() == startIndicesType.getShape().size()) {
+        if (startIndicesType.getShape()[shape.size() - 1] == 1) {
+          reduceWeightTensorDim = true;
+        } else {
+          replaceWithEmbedding = false;
+          llvm::outs() << "Did not satisfy 3" << "\n";
+          llvm::outs() << "Shape size: " << shape.size() << "\n";
+          llvm::outs() << "StartIndicesType shape size: "
+                       << startIndicesType.getShape().size() << "\n";
+        }
+      }
+    });
+
+    llvm::outs() << "replaceWithEmbedding: " << replaceWithEmbedding << "\n";
+    if (replaceWithEmbedding && mlir::isa<mlir::tt::ttir::GatherOp>(gatherOp)) {
```

Comment:
The code as it is written will only replace the last gather op that is found.

Reply:
I was trying to use `walk` but don't quite understand it. I might ping you later to get an understanding of it but I won't use it in the end.